test/state/state.cpp

```diff
@@ -0,0 +1,169 @@
+// evmone: Fast Ethereum Virtual Machine implementation
+// Copyright 2022 The evmone Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "state.hpp"
+#include "host.hpp"
+#include <evmone/evmone.h>
+#include <evmone/execution_state.hpp>
+
+namespace stdx
+{
+/// Implementation of std::erase_if from C++20.
+/// Taken from https://en.cppreference.com/w/cpp/container/unordered_map/erase_if.
+template <class Key, class T, class Hash, class KeyEqual, class Alloc, class Pred>
+typename std::unordered_map<Key, T, Hash, KeyEqual, Alloc>::size_type erase_if(
+    std::unordered_map<Key, T, Hash, KeyEqual, Alloc>& c, Pred pred)
+{
+    auto old_size = c.size();
+    for (auto i = c.begin(), last = c.end(); i != last;)
+    {
+        if (pred(*i))
+            i = c.erase(i);
+        else
+            ++i;
+    }
+    return old_size - c.size();
+}
+}  // namespace stdx
+
+namespace evmone::state
+{
+namespace
+{
+int64_t compute_tx_data_cost(evmc_revision rev, bytes_view data) noexcept
+{
+    constexpr int64_t zero_byte_cost = 4;
+    const int64_t nonzero_byte_cost = rev >= EVMC_ISTANBUL ? 16 : 68;
+    int64_t cost = 0;
+    for (const auto b : data)
+        cost += (b == 0) ? zero_byte_cost : nonzero_byte_cost;
+    return cost;
+}
+
+int64_t compute_access_list_cost(const AccessList& access_list) noexcept
+{
+    static constexpr auto storage_key_cost = 1900;
+    static constexpr auto address_cost = 2400;
+
+    int64_t cost = 0;
+    for (const auto& a : access_list)
+        cost += address_cost + static_cast<int64_t>(a.second.size()) * storage_key_cost;
+    return cost;
+}
+
+int64_t compute_tx_intrinsic_cost(evmc_revision rev, const Transaction& tx) noexcept
+{
+    static constexpr auto call_tx_cost = 21000;
+    static constexpr auto create_tx_cost = 53000;
+    const bool is_create = !tx.to.has_value();
+    const auto tx_cost = is_create && rev >= EVMC_HOMESTEAD ? create_tx_cost : call_tx_cost;
+    return tx_cost + compute_tx_data_cost(rev, tx.data) + compute_access_list_cost(tx.access_list);
+}
+
+evmc_message build_message(const Transaction& tx, int64_t execution_gas_limit) noexcept
+{
+    const auto recipient = tx.to.has_value() ? *tx.to : evmc::address{};
+    return {
+        tx.to.has_value() ? EVMC_CALL : EVMC_CREATE,
+        0,
+        0,
+        execution_gas_limit,
+        recipient,
+        tx.sender,
+        tx.data.data(),
+        tx.data.size(),
+        intx::be::store<evmc::uint256be>(tx.value),
+        {},
+        recipient,
+    };
+}
+}  // namespace
+
+std::optional<std::vector<Log>> transition(
+    State& state, const BlockInfo& block, const Transaction& tx, evmc_revision rev, evmc::VM& vm)
+{
+    if (rev < EVMC_LONDON && tx.kind == Transaction::Kind::eip1559)
```

Comment:
Maybe makes sense to extract transaction-related checks into a separate transaction validity funtion.